File: lib/formats/arc/xp3_archive.rb
require 'lib/binary_io'
require 'lib/memory_file'
require 'zlib'
require_relative 'xp3_archive/filter_factory'

# XP3 archive
# Engine: Kirikiri2
# Extension: .xp3
# Known games:
# - Fate/Stay Night
# - Fate/Hollow Ataraxia
# - Sono Hanabira ni Kuchizuke o 12
module Xp3Archive
  module_function

  MAGIC = "XP3\r\n\x20\x0a\x1a\x8b\x67\x01".b
  FILE_MAGIC = 'File'
  ADLR_MAGIC = 'adlr'
  INFO_MAGIC = 'info'
  SEGM_MAGIC = 'segm'

  def add_cli_help(arg_parser)
    arg_parser.add_help(
      '--plugin=PLUGIN',
      'Selects XP3 decryption routine.',
      possible_values: XP3_FILTERS.keys)
  end

  def parse_cli_options(arg_parser, options)
    filter = arg_parser.switch(['--plugin'])
    filter = filter.nil? ? :none : filter.to_sym
    options[:filter] = filter
  end

  def get_filter(symbol)
    filter = XP3_FILTERS[symbol]
    fail RecognitionError, 'Unknown filter' if filter.nil?
    filter.call
  end

  class Unpacker
    def unpack(arc_file, output_files, options)
      magic = arc_file.read(MAGIC.length)
      fail RecognitionError, 'Not an XP3 archive' unless magic == MAGIC

      version = arc_file.peek(19) { arc_file.read(4) == "\x01\0\0\0" ? 2 : 1 }

      if version == 1
        table_origin = arc_file.read(8).unpack('Q')[0]
      else
        additional_header_offset,
        minor_version = arc_file.read(12).unpack('QI')
        if minor_version != 1
          fail \
            RecognitionError,
            format('Unexpected XP3 version: %s', minor_version)
        end

        arc_file.peek(additional_header_offset) do
          _flag,
          _table_size,
          table_origin = arc_file.read(17).unpack('BQQ')
        end
      end

      table = arc_file.peek(table_origin) { read_raw_table!(arc_file) }
      table = BinaryIO.from_string(table)

      filter = Xp3Archive.get_filter(options[:filter])
      output_files.write { read_file(table, arc_file, filter) } until table.eof?
    end

    private

    def read_raw_table!(arc_file)
      use_zlib = arc_file.read(1).unpack('C')[0] > 0

      if use_zlib
        table_size_compressed,
        table_size_original = arc_file.read(16).unpack('Q<Q<')
        raw = arc_file.read(table_size_compressed)
        raw = Zlib.inflate(raw) if table_size_original != table_size_compressed
        unless raw.length == table_size_original
          fail RecognitionError, 'Bad table size'
        end
        return raw
      end

      raw_size = arc_file.read(8).unpack('Q')[0]
      arc_file.read(raw_size)
    end

    def read_file(raw_table, arc_file, filter)
      magic = raw_table.read(FILE_MAGIC.length)
      fail RecognitionError, 'Expected file chunk' unless magic == FILE_MAGIC

      raw_size = raw_table.read(8).unpack('Q<')[0]
      raw_file_chunk = BinaryIO.from_string(raw_table.read(raw_size))

      info_chunk = Xp3InfoChunk.new
      info_chunk.read!(raw_file_chunk)
      segm_chunks = Xp3SegmChunk.read_list!(raw_file_chunk)
      adlr_chunk = Xp3AdlrChunk.new
      adlr_chunk.read!(raw_file_chunk)

      data = segm_chunks.map { |segm| segm.read_data!(arc_file) } * ''
      data = filter.filter(data, adlr_chunk.encryption_key[0])

      MemoryFile.new(info_chunk.file_name, data)
    end

    # Xp3 SEGM chunk
    class Xp3SegmChunk
      def self.read_list!(arc_file)
        magic = arc_file.read(SEGM_MAGIC.length)
        fail RecognitionError, 'Expected segment chunk' if magic != SEGM_MAGIC

        raw_size = arc_file.read(8).unpack('Q<')[0]
        raw = BinaryIO.from_string(arc_file.read(raw_size))

        chunks = []
        until raw.eof?
          chunk = Xp3SegmChunk.new
          chunk.read!(raw)
          chunks.push(chunk)
        end
        chunks
      end

      def read!(arc_file)
        @flags,
        @offset,
        @original_size,
        @compressed_size = arc_file.read(28).unpack('L<Q<Q<Q<')
      end

      def read_data!(arc_file)
        arc_file.seek(@offset)
        use_zlib = @flags & 7 == 1
        if use_zlib
          raw = Zlib.inflate(arc_file.read(@compressed_size))
          fail RecognitionError, 'Bad SEGM size' if raw.length != @original_size
          return raw
        end

        arc_file.read(@original_size)
      end
    end

    # XP3 INFO chunk
    class Xp3InfoChunk
      attr_accessor :protect
      attr_accessor :original_file_size
      attr_accessor :compressed_file_size
      attr_accessor :file_name

      def read!(arc_file)
        magic = arc_file.read(INFO_MAGIC.length)
        fail RecognitionError, 'Expected info chunk' unless magic == INFO_MAGIC

        raw_size = arc_file.read(8).unpack('Q<')[0]
        raw = BinaryIO.from_string(arc_file.read(raw_size))

        @protect,
        @original_file_size,
        @compressed_file_size,
        name_length = raw.read(22).unpack('I<Q<Q<S<')

        @file_name =
          raw
          .read(name_length * 2)
          .force_encoding('utf-16le')
          .encode('utf-8')
      end
    end

    # XP3 ADLR chunk
    class Xp3AdlrChunk
      attr_accessor :encryption_key

      def read!(arc_file)
        magic = arc_file.read(ADLR_MAGIC.length)
        fail RecognitionError, 'Expected ADLR chunk' unless magic == ADLR_MAGIC

        raw_size = arc_file.read(8).unpack('Q<')[0]
        raw = BinaryIO.from_string(arc_file.read(raw_size))

        @encryption_key = raw.read(4).unpack('L<')
      end
    end
  end

  class Packer
    def pack(arc_file, input_files, options)
      arc_file.write(MAGIC)
      table = prepare_table(input_files, options)
      header_pos = arc_file.tell
      write_dummy_header(arc_file)
      write_contents(arc_file, input_files, table, options)
      table_offset = arc_file.tell
      arc_file.peek(header_pos) { write_header(arc_file, table_offset) }
      write_table(arc_file, table, options)
    end

    private

    def prepare_table(input_files, options)
      table = {}
      origin = 0
      input_files.each do |file|
        e = {
          name: file.name,
          size_original: file.data.length,
          origin: origin + 19,
          random: rand(0xffff_ffff)
        }
        if options[:compress_files]
          e[:size_compressed] = Zlib.deflate(file.data).length
        else
          e[:size_compressed] = file.data.length
        end
        origin += e[:size_compressed]
        table[file.name] = e
      end
      table
    end

    def write_dummy_header(arc_file)
      arc_file.write("\x00" * 8)
    end

    def write_header(arc_file, table_offset)
      arc_file.write([table_offset].pack('Q<'))
    end

    def write_contents(arc_file, input_files, table, options)
      filter = Xp3Archive.get_filter(options[:filter])
      input_files.each do |file|
        file.data = Zlib.deflate(file.data) if options[:compress_files]
        file.data = filter.filter(file.data, table[file.name][:random])
        arc_file.write(file.data)
      end
    end

    def write_table(arc_file, table, options)
      use_zlib = options[:compress_table]

      raw_table = BinaryIO.from_string('')
      table.each do |_name, e|
        info_chunk = prepare_info_chunk(e)
        segm_chunk = prepare_segm_chunk(e, options)
        adlr_chunk = prepare_adlr_chunk(e)
        file_chunk = prepare_file_chunk(info_chunk, segm_chunk, adlr_chunk)
        raw_table.write(file_chunk)
      end

      raw_table.rewind
      raw_table = raw_table.read
      table_size_original = raw_table.length
      raw_table = Zlib.deflate(raw_table) if use_zlib
      table_size_compressed = raw_table.length

      if use_zlib
        arc_file.write("\x01")
        arc_file.write([table_size_compressed].pack('Q<'))
      else
        arc_file.write("\x00")
      end

      arc_file.write([table_size_original].pack('Q<'))
      arc_file.write(raw_table)
    end

    def prepare_file_chunk(info_chunk, segm_chunk, adlr_chunk)
      prepare_chunk(FILE_MAGIC, info_chunk + segm_chunk + adlr_chunk)
    end

    def prepare_info_chunk(table_entry)
      prepare_chunk(INFO_MAGIC, [
        0, # protected?
        table_entry[:size_original],
        table_entry[:size_compressed],
        table_entry[:name].length
      ].pack('I<Q<Q<S<') + table_entry[:name].encode('utf-16le').b)
    end

    def prepare_segm_chunk(table_entry, options)
      prepare_chunk(SEGM_MAGIC, [
        options[:compress_files] ? 1 : 0,
        table_entry[:origin],
        table_entry[:size_original],
        table_entry[:size_compressed]
      ].pack('L<Q<3'))
    end

    def prepare_adlr_chunk(table_entry)
      prepare_chunk(ADLR_MAGIC, [table_entry[:random]].pack('L<'))
    end

    def prepare_chunk(magic, data)
      magic + [data.length].pack('Q<') + data
    end
  end
end

File: src/formats/arc/xp3_archive.cc
// XP3 archive
//
// Company:   -
// Engine:    Kirikiri
// Extension: .xp3
//
// Known games:
// - Fate/Stay Night
// - Fate/Hollow Ataraxia
// - Sono Hanabira ni Kuchizuke o 12
// - Sharin no Kuni, Himawari no Shoujo
// - Comyu Kuroi Ryuu to Yasashii Oukoku

#include <cassert>
#include "buffered_io.h"
#include "formats/arc/xp3_archive.h"
#include "formats/arc/xp3_archive/xp3_filter.h"
#include "formats/arc/xp3_archive/xp3_filter_comyu.h"
#include "formats/arc/xp3_archive/xp3_filter_fha.h"
#include "formats/arc/xp3_archive/xp3_filter_fsn.h"
#include "formats/arc/xp3_archive/xp3_filter_noop.h"
#include "formats/gfx/tlg_converter.h"
#include "string_ex.h"

namespace
{
    const std::string xp3_magic("XP3\r\n\x20\x0a\x1a\x8b\x67\x01", 11);
    const std::string file_magic("File", 4);
    const std::string adlr_magic("adlr", 4);
    const std::string info_magic("info", 4);
    const std::string segm_magic("segm", 4);

    int detect_version(IO &arc_io)
    {
        int version = 1;
        size_t old_pos = arc_io.tell();
        arc_io.seek(19);
        if (arc_io.read_u32_le() == 1)
            version = 2;
        arc_io.seek(old_pos);
        return version;
    }

    bool check_magic(IO &arc_io, std::string expected_magic)
    {
        return arc_io.read(expected_magic.size()) == expected_magic;
    }

    uint64_t get_table_offset(IO &arc_io, int version)
    {
        if (version == 1)
            return arc_io.read_u64_le();

        uint64_t additional_header_offset = arc_io.read_u64_le();
        uint32_t minor_version = arc_io.read_u32_le();
        if (minor_version != 1)
            throw std::runtime_error("Unexpected XP3 version");

        arc_io.seek(static_cast<size_t>(additional_header_offset));
        arc_io.skip(1); // flags?
        arc_io.skip(8); // table size
        return arc_io.read_u64_le();
    }

    std::unique_ptr<IO> read_raw_table(IO &arc_io)
    {
        bool use_zlib = arc_io.read_u8() != 0;
        const uint64_t size_compressed = arc_io.read_u64_le();
        const uint64_t size_original = use_zlib
            ? arc_io.read_u64_le()
            : size_compressed;

        std::string compressed
            = arc_io.read(static_cast<size_t>(size_compressed));
        if (use_zlib)
        {
            std::string uncompressed = zlib_inflate(compressed);
            return std::unique_ptr<IO>(new BufferedIO(uncompressed));
        }
        return std::unique_ptr<IO>(new BufferedIO(compressed));
    }

    void read_info_chunk(IO &table_io, File &target_file)
    {
        if (!check_magic(table_io, info_magic))
            throw std::runtime_error("Expected INFO chunk");
        uint64_t info_chunk_size = table_io.read_u64_le();

        uint32_t info_flags = table_io.read_u32_le();
        uint64_t file_size_original = table_io.read_u64_le();
        uint64_t file_size_compressed = table_io.read_u64_le();

        size_t name_length = table_io.read_u16_le();
        std::string name_utf16 = table_io.read(name_length * 2);
        target_file.name = convert_encoding(name_utf16, "UTF-16LE", "UTF-8");
        assert(info_chunk_size == name_length * 2 + 22);
    }

    bool read_segm_chunk(
        IO &table_io,
        IO &arc_io,
        File &target_file)
    {
        if (!check_magic(table_io, segm_magic))
        {
            table_io.skip(-(signed)segm_magic.size());
            return false;
        }
        uint64_t segm_chunk_size = table_io.read_u64_le();
        assert(segm_chunk_size == 28);

        uint32_t segm_flags = table_io.read_u32_le();
        uint64_t data_offset = table_io.read_u64_le();
        uint64_t data_size_original = table_io.read_u64_le();
        uint64_t data_size_compressed = table_io.read_u64_le();
        arc_io.seek(static_cast<size_t>(data_offset));

        bool use_zlib = (segm_flags & 7) > 0;
        if (use_zlib)
        {
            std::string data_compressed = arc_io.read(
                static_cast<int>(data_size_compressed));
            std::string data_uncompressed = zlib_inflate(data_compressed);
            target_file.io.write(data_uncompressed);
        }
        else
        {
            target_file.io.write_from_io(
                arc_io, static_cast<size_t>(data_size_original));
        }

        return true;
    }

    uint32_t read_adlr_chunk(IO &table_io, uint32_t *encryption_key)
    {
        assert(encryption_key != nullptr);

        if (!check_magic(table_io, adlr_magic))
            throw std::runtime_error("Expected ADLR chunk");

        uint64_t adlr_chunk_size = table_io.read_u64_le();
        assert(adlr_chunk_size == 4);

        *encryption_key = table_io.read_u32_le();
        return true;
    }

    std::unique_ptr<File> read_file(
        IO &arc_io,
        IO &table_io,
        TlgConverter &tlg_converter,
        Xp3Filter *filter)
    {
        std::unique_ptr<File> target_file(new File());

        if (!check_magic(table_io, file_magic))
            throw std::runtime_error("Expected FILE chunk");

        uint64_t file_chunk_size = table_io.read_u64_le();
        size_t file_chunk_start_offset = table_io.tell();

        read_info_chunk(table_io, *target_file);

        while (read_segm_chunk(table_io, arc_io, *target_file))
        {
        }

        uint32_t encryption_key;
        read_adlr_chunk(table_io, &encryption_key);

        assert(table_io.tell() - file_chunk_start_offset == file_chunk_size);

        if (filter != nullptr)
            filter->decode(*target_file, encryption_key);

        tlg_converter.try_decode(*target_file);

        return target_file;
    }
}

struct Xp3Archive::Internals
{
    TlgConverter tlg_converter;
    std::unique_ptr<Xp3Filter> filter;

    Internals() : filter(nullptr)
    {
    }

    ~Internals()
    {
    }
};

Xp3Archive::Xp3Archive() : internals(new Internals)
{
}

Xp3Archive::~Xp3Archive()
{
}

void Xp3Archive::add_cli_help(ArgParser &arg_parser) const
{
    arg_parser.add_help(
        "--plugin=PLUGIN",
        "Selects XP3 decryption routine.\n"
            "Possible values:\n"
            "- comyu\n"
            "- fha\n"
            "- fsn\n"
            "- noop (for unecrypted games)");
    internals->tlg_converter.add_cli_help(arg_parser);
}

void Xp3Archive::parse_cli_options(ArgParser &arg_parser)
{
    const std::string plugin = arg_parser.get_switch("plugin").c_str();
    if (plugin == "comyu")
        internals->filter.reset(new Xp3FilterComyu);
    else if (plugin == "fha")
        internals->filter.reset(new Xp3FilterFha);
    else if (plugin == "fsn")
        internals->filter.reset(new Xp3FilterFsn);
    else if (plugin == "noop")
        internals->filter.reset(new Xp3FilterNoop);
    else
        throw std::runtime_error("Unrecognized plugin: " + plugin);

    internals->tlg_converter.parse_cli_options(arg_parser);
}

void Xp3Archive::unpack_internal(File &file, FileSaver &file_saver) const
{
    if (!check_magic(file.io, xp3_magic))
        throw std::runtime_error("Not an XP3 archive");

    int version = detect_version(file.io);
    uint64_t table_offset = get_table_offset(file.io, version);
    file.io.seek((uint32_t)table_offset);
    std::unique_ptr<IO> table_io = read_raw_table(file.io);

    while (table_io->tell() < table_io->size())
    {
        file_saver.save(read_file(
            file.io,
            *table_io,
            internals->tlg_converter,
            internals->filter.get()));
    }
}
